contracts/Strategy.sol

```diff
@@ -53,80 +222,620 @@ contract Strategy is BaseStrategy {
             uint256 _debtPayment
         )
     {
-        // TODO: Do stuff here to free up any returns back into `want`
-        // NOTE: Return `_profit` which is value generated by all positions, priced in `want`
-        // NOTE: Should try to free up at least `_debtOutstanding` of underlying position
+        for (uint256 i = 0; i < kashiPairs.length; i++) {
+            accrueInterest(i);
+            depositKashiInMasterChef(i); // claim and deposit loose
+        }
+
+        sell();
+
+        uint256 assets = estimatedTotalAssets();
+        uint256 wantBal = balanceOfWant();
+
+        uint256 debt = vault.strategies(address(this)).totalDebt;
+
+        if (assets >= debt) {
+            _profit = assets.sub(debt);
+        } else {
+            _loss = debt.sub(assets);
+        }
+
+        _debtPayment = _debtOutstanding;
+        uint256 amountToFree = _debtPayment.add(_profit);
+
+        if (amountToFree > 0 && wantBal < amountToFree) {
+            liquidatePosition(amountToFree.sub(wantBal));
+
+            uint256 newLoose = balanceOfWant();
+
+            // if we didnt free enough money, prioritize paying down debt before taking profit
+            if (newLoose < amountToFree) {
+                if (newLoose <= _debtPayment) {
+                    _profit = 0;
+                    _debtPayment = newLoose;
```

Comment:
shouldn't you add a loss here? 
if you tried to repay debt and for some reason you got less, maybe we need to increment loss?

just a thought, not 100% sure

Reply:
I think you're correct

Reply:
We will take some losses when the kashiPair utilization is high and as a result can't free, but that's probably okay.